### PR TITLE
Use cm.getModeAt in TokenUtils.getModeAt

### PR DESCRIPTION
**Pull request by marcelgerber**
So, `cm.getModeAt` basically just does what our function did, but faster. See https://github.com/codemirror/CodeMirror/blob/b833e63142d18e89a43cc37b712325f5b530c08c/lib/codemirror.js#L4228-L4232

### CODE REVIEW

File: src/utils/TokenUtils.js [hunk 1]
```diff
@@ -211,14 +210,10 @@ define(function (require, exports, module) {
      * @return {mode:{Object}, name:string}
      */
     function getModeAt(cm, pos) {
-        var outerMode = cm.getMode(),
-            modeData = CodeMirror.innerMode(outerMode, cm.getTokenAt(pos, true).state),
```
Comment (ingorichter) [inline src/utils/TokenUtils.js]: The only thing that I don't understand, is what happened to the second argument of `getTokenAt`. This seemed to be something that cm refers to as `precision`. This is now handled differently in cm. It seems, following `cm.getModeAt` code path, it will be undefined. Perhaps this doesn't make any difference, but I don't know of a good way to proof it. Perhaps @RaymondLim knows more about the internal mechanics of cm.

Comment (MarcelGerber) [inline src/utils/TokenUtils.js]: Yeah, you're right, `precise` defaults to `false`.
Apparently, `precise` only affects [`findStartLine`](https://github.com/codemirror/CodeMirror/blob/3cf2e251a3037089a626f53facfe91bfe65d790b/lib/codemirror.js#L1558-L1562), but I don't know to what extent.

**Comment by ingorichter** [comment]
@RaymondLim do you know about the `precision` parameter and what the impact might be, when we pass `undefined` vs. `true`?

**Comment by RaymondLim** [comment]
@ingorichter I don't know about the `precision` param and its impact, but as @MarcelGerber pointed out it only affects `findStartLine` function. So I guess this pr won't affect in most of the times except in an edge case where we need to search backwards passing 100 lines (1000 lines if the current pos is inside the inner mode). I can't imagine how to create a valid test case like that since searching backwards may stop as soon as `line.stateAfter` is true in the following code.

```
  function findStartLine(cm, n, precise) {
    var minindent, minline, doc = cm.doc;
    var lim = precise ? -1 : n - (cm.doc.mode.innerMode ? 1000 : 100);
    for (var search = n; search > lim; --search) {
      if (search <= doc.first) return doc.first;
      var line = getLine(doc, search - 1);
      if (line.stateAfter && (!precise || search <= doc.frontier)) return search;
      var indented = countColumn(line.text, null, cm.options.tabSize);
      if (minline == null || minindent > indented) {
        minline = search - 1;
        minindent = indented;
      }
    }
    return minline;
  }
```

**Comment by ingorichter** [comment]
@MarcelGerber @RaymondLim if this is a rather unlikely scenario, then the impact might be minimal on a daily basis. Thanks for the info Raymond.

**Comment by MarcelGerber** [comment]
The [CM manual](http://codemirror.net/doc/manual.html#getTokenAt) says:
> If `precise` is true, the token will be guaranteed to be accurate based on recent edits. If false or not specified, the token will use cached state information, which will be faster but might not be accurate if edits were recently made and highlighting has not yet completed.

So if I get this right, in most cases it doesn't matter whether it's set or not.
I'd like to know though if this only applies to big changes (like copy-paste-replacing a whole file) or also for small ones.

**Comment by MarcelGerber** [comment]
Closing in favor of #10882, which improves `TokenUtils.getModeAt` to be a lot faster in cases.
